### PR TITLE
Add IPv6 support

### PR DESCRIPTION
**Pull request by ab77**
* remove IP query from Supervisor API
* add IPv4 address query function
* add IPv6 address query function
* register both IPv4 and IPv6 addresses

Update open-balena-base from 11.3.9 to 11.3.12

Change-type: patch

### CODE REVIEW

File: src/app.ts [hunk 1]
```diff
@@ -65,62 +34,36 @@ const dbusInvoker = (message: Message): PromiseLike<any> => {
...
+): Array<os.NetworkInterfaceInfo | undefined> => {
+	if (iface) {
+		return Object.values([os.networkInterfaces()[iface]]).flatMap((infos) => {
+			return infos?.find((i) => i.family === 'IPv4' && !i.internal);
+		});
+	} else {
+		return Object.values(os.networkInterfaces()).flatMap((infos) => {
+			return infos?.find((i) => i.family === 'IPv4' && !i.internal);
+		});
 	}
```
Comment (Hades32) [inline src/app.ts:48]: If you put everything in one chain, the undefined in the type should go away
```suggestion
	Object.entries(os.networkInterfaces())
	  .filter(([nic,ips]) => !iface || nic === iface)
	  .flatMap(([,ips])=>ips)
	  .filter(ip=>!ip.internal && ip.family === "IPv4")
```

File: src/app.ts [hunk 2]
```diff
@@ -267,25 +214,36 @@ const balena = BalenaSdk({
...
 		if (process.env.INTERFACE) {
-			ipAddr = getNamedInterfaceAddr(process.env.INTERFACE);
+			// get the first non link-local IP address for the specified interface for each address family
+			ipAddrs.push(
+				getAddrs4(process.env.INTERFACE).flatMap((f) =>
+					f ? f.address : [],
+				)[0],
+				getAddrs6(process.env.INTERFACE).flatMap((f) =>
+					f ? f.address : [],
+				)[0],
```
Comment (Hades32) [inline src/app.ts:226]: there could be interfaces with only local addresses, so compacting makes sense and let's us get rid of the extra flatMap
```suggestion
				..._.compact( [
				getAddrs4(process.env.INTERFACE)[0],
				getAddrs6(process.env.INTERFACE)[0]
				]),
```

Comment (Hades32) [reply src/app.ts:226]: similar change needed below of course

File: src/app.ts (hunk 1 above)
Comment (Hades32) [reply src/app.ts:48]: similar change needed below of course

Comment (ab77) [reply src/app.ts:48]: @Hades32 so something like:

```
const getAddrs6 = (
	iface?: string,
): Array<os.NetworkInterfaceInfo | undefined> => {
	return Object.entries(os.networkInterfaces())
	  .filter(([nic,ips]) => !iface || nic === iface)
	  .flatMap(([,ips])=>ips)
	  .filter(ips=>!ip.internal && ip.family === "IPv6" && ip.scopeid === 0)
};
```

Comment (Hades32) [reply src/app.ts:48]: yes, but we don't need the ` | undefined` anymore

Comment (ab77) [reply src/app.ts:48]: @Hades32 it's angry...

```
> @1.9.14 build /Users/anton/dev/balena-io/balena-mdns-publisher
> tsc
                                                                    
src/app.ts:40:2 - error TS2322: Type '(NetworkInterfaceInfoIPv4 | NetworkInterfaceInfoIPv6 | undefined)[]' is not assignable to type 'Ne
tworkInterfaceInfo[]'.                                                                                                                  
  Type 'NetworkInterfaceInfoIPv4 | NetworkInterfaceInfoIPv6 | undefined' is not assignable to type 'NetworkInterfaceInfo'.
    Type 'undefined' is not assignable to type 'NetworkInterfaceInfo'.
                                                                    
40  return Object.entries(os.networkInterfaces())
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~                                                                                       
41    .filter(([nic,ips]) => !iface || nic === iface)
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
42    .flatMap(([,ips])=>ips)                                       
   ~~~~~~~~~~~~~~~~~~~~~~~~~~
43    .filter(ip=>!ip.internal && ip.family === "IPv4")                                                                                 
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~     

src/app.ts:41:18 - error TS6133: 'ips' is declared but its value is never read.                   
                                                                                                                                        
41    .filter(([nic,ips]) => !iface || nic === iface)
                    ~~~                                                                                                                 
                                                                                                                                        
src/app.ts:43:17 - error TS2532: Object is possibly 'undefined'.

43    .filter(ip=>!ip.internal && ip.family === "IPv4")
                   ~~
                                  
src/app.ts:43:32 - error TS2532: Object is possibly 'undefined'.
                                  
43    .filter(ip=>!ip.internal && ip.family === "IPv4")
                                  ~~
...
```

Comment (Hades32) [reply src/app.ts:48]: @ab77 seems it's valid to have IPs at all. We can just replace that case with an empty array:
```typescript
let ips = Object.entries(os.networkInterfaces())
      .filter(([nic]) => (!iface || nic === iface))
      .flatMap(([, ips]) => ips || [])
      .filter((ip) => !ip.internal && ip.family === 'IPv4');
```

Comment (ab77) [reply src/app.ts:48]: @Hades32 that got rid of most of the red, only have this now:

```
src/app.ts:39:18 - error TS6133: 'ips' is declared but its value is never read.

39      .filter(([nic, ips]) => !iface || nic === iface)
                       ~~~
```

I tried `noUnusedLocals: true|false` in tsconfig, doesn't make any difference...

Comment (dfunckt) [reply src/app.ts:48]: You should be able to just skip the argument or mark it as ignored by prefixing it with an underscore. Don't mess with tsconfig -- it's (usually) set up sensibly.

Option A (preferred):

```
.filter(([nic]) => !iface || nic === iface)
```

Option B:

```
.filter(([nic, _ips]) => !iface || nic === iface)
```

Comment (ab77) [reply src/app.ts]: @dfunckt that was easy, thanks.

**Comment by ab77** [comment]
@balena-ci I self-certify!

**Pull request merged**